crates/fuel-core/src/graphql_api/da_compression.rs

```diff
@@ -0,0 +1,212 @@
+                    .replace(key, value)?;
+
+                // Remove the overwritten value from index, if any
+                if let Some(old_value) = old_value {
+                    let old_reverse_key = (&old_value).into();
+                    self.db_tx
+                        .storage_as_mut::<DaCompressionTemporalRegistryIndex>()
+                        .remove(&old_reverse_key)?;
+                }
+
+                // Add the new value to the index
+                let reverse_key = value.into();
+                self.db_tx
+                    .storage_as_mut::<DaCompressionTemporalRegistryIndex>()
+                    .insert(&reverse_key, key)?;
+
+                // Update the timestamp
+                self.db_tx
+                    .storage_as_mut::<DaCompressionTemporalRegistryTimestamps>()
+                    .insert(&TimestampKey { keyspace: TimestampKeyspace::$type, key: *key }, &timestamp)?;
+
+                Ok(())
+            }
+
+            fn registry_index_lookup(
+                &self,
+                value: &$type,
+            ) -> anyhow::Result<Option<fuel_core_types::fuel_compression::RegistryKey>>
+            {
+                let reverse_key = value.into();
+                Ok(self
+                    .db_tx
+                    .storage_as_ref::<DaCompressionTemporalRegistryIndex>()
+                    .get(&reverse_key)?
+                    .map(|v| v.into_owned()))
+            }
+        }
+
+        impl<'a, Tx> EvictorDb<$type> for CompressTx<'a, Tx>
+        where
+            Tx: OffChainDatabaseTransaction,
+        {
+            fn set_latest_assigned_key(
+                &mut self,
+                key: fuel_core_types::fuel_compression::RegistryKey,
+            ) -> anyhow::Result<()> {
+                self.db_tx
+                    .storage_as_mut::<DaCompressionTemporalRegistryEvictorCache>()
+                    .insert(&MetadataKey::$type, &key)?;
+                Ok(())
+            }
+
+            fn get_latest_assigned_key(
+                &self,
+            ) -> anyhow::Result<Option<fuel_core_types::fuel_compression::RegistryKey>> {
+                Ok(self
+                    .db_tx
+                    .storage_as_ref::<DaCompressionTemporalRegistryEvictorCache>()
+                    .get(&MetadataKey::$type)?
+                    .map(|v| v.into_owned())
+                )
+            }
+        }
+
+    }};
+}
+
+impl_temporal_registry!(Address);
+impl_temporal_registry!(AssetId);
+impl_temporal_registry!(ContractId);
+impl_temporal_registry!(ScriptCode);
+impl_temporal_registry!(PredicateCode);
+
+impl<'a, Tx> UtxoIdToPointer for CompressTx<'a, Tx>
+where
+    Tx: OffChainDatabaseTransaction,
+{
+    fn lookup(
+        &self,
+        utxo_id: fuel_core_types::fuel_tx::UtxoId,
+    ) -> anyhow::Result<fuel_core_types::fuel_tx::CompressedUtxoId> {
+        for event in self.block_events {
```

Comment:
If we have many events in the transaction is can be really long loop. It would be nice to index all events on construction and store in the hash map.

Reply:
Agreed. Follow-up: https://github.com/FuelLabs/fuel-core/issues/2279